Integrate with detekt-hint

Mkohm opened the pull request:
I'll try again... Danger is having some trouble with PRs from forks (https://github.com/danger/danger/issues/1103).

As discussed in #2437, here is a PR with the integration of detekt-hint.

I would be very happy if you find any value of it - please let me know if there are any problems or required changes. If you have any feedback, please let me know!

arturbosch commented on .github/workflows/pre-merge.yaml:
```diff
@@ -76,6 +76,13 @@ jobs:
+      uses: actions/setup-ruby@v1.0.0
+
+    - name: Run Danger
+      run: gem install bundler && bundle install && bundle update danger && bundle exec danger --verbose
```
This will run on every matrix entry. 
detekt-hint could be an own action so it is run just once

arturbosch commented on Gemfile:
```diff
@@ -0,0 +1,7 @@
+
+gem 'danger'
+gem 'oga'
+gem "danger-kotlin_detekt", "~> 0.0.3"
```
Hm, the ruby part looks really generic. 
Could this be a custom GitHub action which bundles the ruby action and the whole danger thing?
We are not a ruby application and I would really like to avoid extra dependencies.

Mkohm replied:
I did'nt know, that is a really good idea! Looks like i can create a custom Github action that runs within a docker container, and put the ruby stuff and Danger file in there.

arturbosch commented on build.gradle.kts:
```diff
@@ -358,6 +359,24 @@ val detektAll by tasks.registering(Detekt::class) {
     }
 }
 
+task<Detekt>("detektHint") {
```
This task is created eagerly. Please use `tasks.register` to make it lazy.

arturbosch commented on build.gradle.kts:
```diff
@@ -358,6 +359,24 @@ val detektAll by tasks.registering(Detekt::class) {
+task<Detekt>("detektHint") {
+    description = "Running detekt-hint and outputs report for Danger to consume."
+    setSource(analysisDir)
+    config.setFrom(file("$rootDir/config/detekt/detekt-hint.yml"))
```
Lets add here a `disableDefaultRuleSets` so only detekt-hint gets executed.

arturbosch commented on config/detekt/detekt-hint.yml:
```diff
@@ -0,0 +1,11 @@
+detekt-hint:
+  UseCompositionInsteadOfInheritance:
+    active: true
+    yourUniquePackageName: "io.gitlab.arturbosch"
```
It should be save to detekt as the fourth package.

Mkohm commented:
> Edit: It seems that I need to register somewhere and provide a GitHub token .. ?

If you want a separate bot only for detekt, you need to add a token to the secrets in the detekt-github repository. Downside of doing that is that pull requests from forks will not be able to get comments on their PR's by Danger. To support PR's from forks (which i guess we want to) we can use the "GITHUB_TOKEN"  which is created by Github.  

However, i think the issue in Danger (https://github.com/danger/danger/issues/1103) needs to be resolved for this to work.

Mkohm commented:
I have addressed your comments @arturbosch, i think it works better this way! 

Unfortunately, due to the bug in danger (https://github.com/danger/danger/issues/1103), detekt-hint will not work on pr's from forks. In that case, the action will succeed (even when it fails due to the linked issue). I hope you still want to help me out by testing the tool :)
